### DISCUSSION TITLE
What's the point of Mise over ASDF or Aqua?

### DISCUSSION THREAD

**Discussion by airtonix** [jdx/mise#2441] [Q&A]
So far having used asdf for many years my two main gripes are:

- that it doesn't work on windows.
- it doesn't help me protect against supply chain attacks by: 
  - hash pinning my asdf plugins per project. (asdf-plugin-manager tries to address this but it's broken)
  - hash pinning the tool versions

So after using Mise for a couple weeks : 

I love:

- it's friendly output
- it leans on the asdf plugin ecosystem

One thing that Mise does which I had always made a asdf wrapper for:

- it auto installs asdf plugins

but I now realise there's no real point to it because what we really need is: 

- plugin system that works on bash/zsh/fish/ash/dash/powershell/nushell (basically it can't be based on shell scripting)
- SBOM to protect against supply chain attack
  - asdf-plugin-manager tried to do this, but there's a few issues with it.
  - asdf plugins themselves don't universally support gtisha, and lean on git tags which can be moved.

I have been keeping an eye on another Tool Manager called [Aqua](https://aquaproj.github.io/) that solves all these problems.

We don't really need a task runner or a env manager. There are existing tools that solve these problems properly.

**Comment by mcallaway** [comment]
I'm pretty new to mise, and this question is how I discovered [aqua](https://aquaproj.github.io/), so I went over there to compare. Before that, however, I just wanted to comment to the original poster that I *do* want a task runner so that I don't need Makefiles, and I do want environment variable control. Looks like Windows support is coming to mise (or is already here, to some level). So perhaps hash pinning is your only missing feature?

By contrast aqua says one [must use asdf for language support](https://aquaproj.github.io/docs/#use-aqua-along-with-asdf) and I'm primarily interested in python. But I don't want two systems (aqua *and* asdf).

So it seems for me, mise is more feature complete. I went to ask [what's the point of aqua](https://aquaproj.github.io/docs/#comparison). Not yet sure what I think overall.

**Reply by airtonix** [reply]
@mcallaway cant we link to the ticket tracking this statement? 

> Looks like Windows support is coming to mise

**Comment by jdx** [comment]
If aqua solves your problems then you should use aqua. I'm not very familiar with the project but my understanding is that it has more of a focus on security and reproducibility where I have more of a focus on ease-of-use and features.

I've moved mise away from being a simple asdf replacement and it now has support for running vfox plugins as well. I have not investigated this but I could see a backend being added to support aqua through mise—perhaps via a mise-aqua plugin. The problems I'm interested in are more to do with the front-end of dev tools over the back-end so leveraging work that others have done in that space may be useful.

**Reply by airtonix** [reply]
The vfox plugin support is a big step forward. ❤️ 

If you tackle the SBOM question, then mise becomes a very strong tool. Supply chain attack is a very real problem that too many people ignore (thus becoming a systemic issue).

**Reply by borissmidt** [reply]
In my opinion it might be a good idea to implement aqua registry support directly into mise and make it the default backend.
It is easier to register a new too into aqua than asdf it is cross platform compatible and its easier to download the tools in parallel.

I had a few 'broken' asdf plugins which required jq so then i had to make sure jq was also in my tool-versions file.

It might not be that much work if this issue gets resolved:
https://github.com/aquaproj/aqua/issues/3074

**Reply by boris-smidt-klarrio** [reply]
@jdx Cool that you implemented the aqua backend :+1: i will try it out today.

**Comment by powerman** [comment]
> We don't really need a task runner or a env manager. There are existing tools that solve these problems properly.

Can you name these tools? I've checked a lot, and first one which actually solve these problems properly is Mise.

From top of my head:
- Whole dotenv family of tools are broken by design just because there are [no common dotenv specification](https://github.com/dotenv-linter/dotenv-linter/discussions/805) and you can't use only one of them (and thus stick to dotenv variant of that tool): a lot of 3rd-party tools use their own libraries to read dotenv, do that by default, do that using different dotenv syntax/features, and this guaranteed to give you unpredictable results sooner or later.
- Taskfile in about 7 years since releasing 1.0.0 didn't solved an issue with buggy and counter-intuitive [precedence for env vars](https://github.com/go-task/task/issues/1038) - which to me looks like a critical issue which I've noticed in a hour when I start playing with Task. A couple of months ago they tried to fix it as [backward-incompatible experiment](https://taskfile.dev/experiments/env-precedence/). 7 years, Karl! How many similar issues are lurking there I just don't like to know, sorry.
- ASDF - well, everyone here know why Mise is much better.
- Aqua… it's mostly looks good - for what it does. But it does not install as much tools as Mise (I'm using more than hundred actually, and Mise installed all but 3, while Aqua installed about 25 or so), it does not implement env vars and tasks. Also it shims are much slower than Mise (50 ms vs 8 ms on my system). The only thing it does really good is security - I hope Mise eventually will have similar features.

**Reply by jdx** [reply]
> Whole dotenv family of tools are broken by design just because there are https://github.com/dotenv-linter/dotenv-linter/discussions/805 and you can't use only one of them (and thus stick to dotenv variant of that tool): a lot of 3rd-party tools use their own libraries to read dotenv, do that by default, do that using different dotenv syntax/features, and this guaranteed to give you unpredictable results sooner or later.

I hadn't considered this but just by putting this in toml and not .env has a lot of benefits by establishing our own rules about the file. xkcd 927 and all that—I don't see that as valid criticism. Put your env vars in toml and it works, that's what matters.

> Taskfile in about 7 years since releasing 1.0.0 didn't solved an issue with buggy and counter-intuitive https://github.com/go-task/task/issues/1038

lol that  is ridiculous. I can't imagine anyone would want it to behave the way it does! If that was my project I'd just declare it a bug and flip the behavior—breaking change be damned.

I suppose the answer to OP's question though is that with me at the helm you are more likely to experience change. asdf hasn't had a new feature in 2 years. While I try to ensure stability, my top priority is making mise the best tool manager it can be and when those goals are at odds I lean towards making mise better.

**Reply by jdx** [reply]
I just discovered just has the same exact behavior as Taskfile: https://github.com/casey/just/issues/733

this is crazy to me. I don't know why anyone would think that's how it should behave. And given these are some of the top issues on both projects—and nobody has asked for this behavior in mise to be different AFAIK—I'm inclined to think I made the right decision with mise.

to me env vars are what was there before you ran the command and since you're running the command, things it sets should overwrite what was there before

**Comment by koppor** [comment]
For me it is very simple: [aqua does not support java](https://github.com/aquaproj/aqua-registry/issues/1044).

**Comment by jakewilliami** [comment]
> ASDF - well, everyone here know why Mise is much better.

@powerman I don't :slightly_smiling_face:  I thought this would be in the FAQ but can't see anywhere where this is discussed.  I have no experience with Mise or ASDF (or Aqua).  If I want to use one of them, how would I know if my use-case is sufficient and which one to use?

P.S., I never had to learn [Make](https://www.gnu.org/software/make/) because [just](https://github.com/casey/just) seems to be sufficient to me.  This is the only build tool I use.  When it comes to CI/CD, someone has usually made a convenient action for my use-case.

**Reply by pboling** [reply]
I used asdf for many years, and recently switched to mise, and it I've had fewer issues with it.  It doesn't rely on shims the way asdf does, and that was the main issue I consistently had with asdf. I've never used aqua, and have no idea why I would want to.  mise + direnv is everything I need.
Update: I no longer use direnv - mise replaces everything I had used direnv for.

**Reply by powerman** [reply]
>  I thought this would be in the FAQ but can't see anywhere where this is discussed.

https://mise.jdx.dev/dev-tools/comparison-to-asdf.html

As for `just` - I don't use it, but at a glance main selling point for mise is support for tools management. It means all developers on a project will use same versions of all tools used on a project. And even if you're working alone on several projects with mise each project will use required versions of all tools instead of single version you've installed globally on your system. I suppose nothing stops you from using `mise` and `just` together - `mise` for tools (including `just`) management and probably setting project environment vars, `just` to run tasks (until you migrate tasks to mise too).

**Reply by jakewilliami** [reply]
> https://mise.jdx.dev/dev-tools/comparison-to-asdf.html

Thanks, I did see that.  I guess what I mean is that it wasn't helpful for me without prior knowledge of `asdf`.

> all developers on a project will use same versions of all tools used on a project.

That sounds really good.  So it's like a package manager but for CLI.

For Python development, I use [`uv`](https://github.com/astral-sh/uv/), and store information about the version of Python in `pyproject.toml`.  However, I can see this being very useful for other projects.

I'm not sure if I see it being as useful for library development?  I suppose it's useful for running different versions locally, but modern toolchain managers have support for running commands at different versions (e.g., [`rustup`](https://rustup.rs/)/[`juliaup`](github.com/JuliaLang/juliaup)).  I think I broadly see the vision, though.  Thanks :slightly_smiling_face:

> until you migrate tasks to mise too

And good to know `mise` is a task runner as well.

**Reply by airtonix** [reply]
Something `just` has over `mise` is that the friction to creating nested commands throughout a monorepo are far less.

with mise i have to essentially replicate the root `.mise/tasks` directory structure through out the monorepo.

it'd be nice if mise would allow us to namespace included tasks.

`justfile`
```justfile

setup:
  # some global monorepo setup and provisioning task

app APPNAME="" COMMAND="dev":
    just \
        --justfile=apps/{{APPNAME}}/justfile \
        --working-directory=apps/{{APPNAME}} \
        {{COMMAND}}
```

`apps/something/justfile`
```justfile
dev:
  npm run dev
```

lets me do `just app:something dev`


Now achieving the same thing with mise is done like this

`.mise/tasks/setup`
```sh
#!/bin/bash

# some global monorepo setup and provisioning task
```


`apps/something/.mise/tasks/apps-something/dev`
```sh
#!/bin/bash

npm run dev
```

and then i need to : 

`.mise.toml`
```
...
[task_config]
includes = [
  ".mise/tasks",
  "apps/something/.mise/tasks",
]
```

which lets me do `mise run apps-something:dev`


but it'd be a lot better if i could just do:


`.mise/tasks/setup`
```sh
#!/bin/bash

# some global monorepo setup and provisioning task
```


`apps/something/.mise/tasks/dev`
```sh
#!/bin/bash

npm run dev
```

`.mise.toml`
```
...
[task_config.includes]
root = ".mise/tasks" # optional since it should always be there/
something = "apps/something/.mise/tasks" # namespaced
```

or even better, just recursively detect `(namespace).mise/tasks` and do the name spacing for me .

**Comment by koppor** [comment]
I am really liking [gg.cmd](https://github.com/eirikb/gg). It is a script running on **both** macOS and Windows (and also Linux) without the need to install anything else.

Besides build-in support for `npx`, ´node`, `just`, etc., it allows to "execute" any GitHub project (based on the releases there).

For instance, to run [GitHub CLI](https://github.com/cli/cli), one just executes following command:

    ./gg.cmd gh/cli/cli

If it comes to Node oder Java, Node itself or the JDK are downloaded on demand - respecting the version requirements of gradle etc.

The concept is described at <https://github.com/eirikb/gg/tree/main/src>.

**Comment by ParadaCarleton** [comment]
If I can ask a more specific question: is there any advantage/disadvantage of the native mise vs. aqua backends? Some of the comments here make it sound like aqua is more secure somehow, but I'm not sure I follow.

**Comment by JamBalaya56562** [comment]
@ParadaCarleton — I measured this rather than reasoning from the docs, and the result cuts against the impression the thread gives.

Installing **the same tool at the same version** through both backends on **v2026.8.3**:

```console
mise github:magefile/mage@1.17.2 [2/3] checksum mage_1.17.2_Linux-64bit.tar.gz
mise github:magefile/mage@1.17.2 [2/3] verify GitHub artifact attestations
mise github:magefile/mage@1.17.2 [2/3] verify SLSA provenance
mise github:magefile/mage@1.17.2 ✓ installed

mise aqua:magefile/mage@1.17.2 [2/3] checksum mage_1.17.2_Linux-64bit.tar.gz
mise aqua:magefile/mage@1.17.2 ✓ installed
```

The `github:` backend verified **more**, not less.

The reason is that **aqua verifies what its registry entry declares**, and mage's entry declares only a checksum — [`pkgs/magefile/mage/registry.yaml`](https://github.com/aquaproj/aqua-registry/blob/main/pkgs/magefile/mage/registry.yaml) has `checksum:` blocks and no `cosign:`, `slsa_provenance:` or `minisign:`. mise honours all of those when they are present — `aqua.cosign`, `aqua.slsa`, `aqua.minisign` and `aqua.github_attestations` all default to `true` — so for a package whose entry carries signature metadata, aqua does more. For one that doesn't, it does less than mise's own backend, which goes looking for attestations and provenance on the release itself.

So the honest answer to "is aqua more secure" is: **it is a property of the individual package entry, not of the backend.** `mise install` prints what it verified, which is the quickest way to check any specific tool.

On @airtonix's original hash-pinning gripe, which is the other half of this: the lockfile is backend-independent. Both installs above wrote the same rows —

```toml
[tools."aqua:magefile/mage"."platforms.linux-x64"]
checksum = "sha256:b1dd189f5a4d38484176dd5be3b651eb7cbc0b78eaf4bb9715738aa24edec644"
url = "https://github.com/magefile/mage/releases/download/v1.17.2/mage_1.17.2_Linux-64bit.tar.gz"
```

— identical checksums and URLs under `github:` and `aqua:` alike, for every platform rather than just the one being installed. So version and artifact pinning does not depend on which backend you pick.
